ably/options.go

```diff
@@ -14,36 +15,86 @@ const (
 	ProtocolMsgPack = "msgpack"
 )
 
+var errInvalidKey = errors.New("invalid key format")
+
+var DefaultOptions = &ClientOptions{
+	RestEndpoint:     "rest.ably.io",
+	RealtimeEndpoint: "realtime.ably.io",
+	Protocol:         ProtocolJSON, // TODO: make it ProtocolMsgPack
+}
+
 type ClientOptions struct {
-	RealtimeEndpoint string
 	RestEndpoint     string
-	Key              string
-	Token            string
-	Secret           string
+	RealtimeEndpoint string
+	KeyName          string
+	KeySecret        string
 	ClientID         string
-	Protocol         string
+	Protocol         string // either ProtocolJSON or ProtocolMsgPack
 	UseTokenAuth     bool
 	NoTLS            bool
 
 	HTTPClient *http.Client
 }
 
-func (p *ClientOptions) Prepare() {
-	if p.Key != "" {
-		p.parseKey()
+func NewClientOptions(key string) (*ClientOptions, error) {
+	opts := &ClientOptions{}
+	if err := opts.SetKey(key); err != nil {
+		return nil, err
+	}
+	return opts, nil
+}
+
+func (opts *ClientOptions) restURL() string {
+	host := opts.RestEndpoint
+	if host == "" {
+		host = DefaultOptions.RestEndpoint
+	}
+	if opts.NoTLS {
+		return "http://" + host
+	}
+	return "https://" + host
+}
+
+func (opts *ClientOptions) realtimeURL() string {
+	host := opts.RealtimeEndpoint
+	if host == "" {
+		host = DefaultOptions.RealtimeEndpoint
+	}
+	if opts.NoTLS {
+		return "ws://" + host + ":80"
+	}
+	return "wss://" + host + ":443"
+}
+
+func (opts *ClientOptions) SetKey(key string) error {
```

Comment:
Ok, well I think we should ideally make Options not a public struct type, and instead provide accessors once the library is instanced so that unless someone intentionally pokes into the library and tries to break things, they will not mistakenly be led to believe they can change attributes.